packages/grid/_modules_/grid/hooks/features/detailPanel/useGridDetailPanel.ts

```diff
@@ -0,0 +1,246 @@
+      if (params.field !== GRID_DETAIL_PANEL_TOGGLE_FIELD || getDetailPanelContent == null) {
+        return;
+      }
+
+      const content = contentCache[params.id];
+      if (!React.isValidElement(content)) {
+        return;
+      }
+
+      // Ignore if the user didn't click specifically in the "i" button
+      if (event.target === event.currentTarget) {
+        return;
+      }
+
+      apiRef.current.toggleDetailPanel(params.id);
+    },
+    [apiRef, contentCache, getDetailPanelContent],
+  );
+
+  const handleCellKeyDown = React.useCallback<GridEventListener<GridEvents.cellKeyDown>>(
+    (params, event) => {
+      if (!event.ctrlKey || event.key !== 'Enter' || props.getDetailPanelContent == null) {
+        return;
+      }
+      apiRef.current.toggleDetailPanel(params.id);
+    },
+    [apiRef, props.getDetailPanelContent],
+  );
+
+  useGridApiEventHandler(apiRef, GridEvents.cellClick, handleCellClick);
+  useGridApiEventHandler(apiRef, GridEvents.cellKeyDown, handleCellKeyDown);
+
+  const addToggleColumn = React.useCallback(
+    (columnsState: GridColumnsRawState) => {
+      if (
+        columnsState.lookup[GRID_DETAIL_PANEL_TOGGLE_FIELD] !== undefined ||
+        props.getDetailPanelContent == null
+      ) {
+        return columnsState;
+      }
+
+      return {
+        ...columnsState,
+        all: [GRID_DETAIL_PANEL_TOGGLE_FIELD, ...columnsState.all],
+        lookup: {
+          [GRID_DETAIL_PANEL_TOGGLE_FIELD]: {
+            ...GRID_DETAIL_PANEL_TOGGLE_COL_DEF,
+          },
+          ...columnsState.lookup,
+        },
+      };
+    },
+    [props.getDetailPanelContent],
+  );
+
+  const addDetailHeight = React.useCallback(
+    (initialValue: Record<string, number>, row: GridRowEntry) => {
+      if (expandedRowIds.length === 0 || !expandedRowIds.includes(row.id)) {
+        return { ...initialValue, detail: 0 };
+      }
+      return {
+        ...initialValue,
+        detail: heightCache[row.id] ?? 0, // Fallback to zero because the cache might not be ready yet (e.g. page was changed)
+      };
+    },
+    [expandedRowIds, heightCache],
+  );
+
+  useGridRegisterPreProcessor(apiRef, GridPreProcessingGroup.hydrateColumns, addToggleColumn);
+  useGridRegisterPreProcessor(apiRef, GridPreProcessingGroup.rowHeight, addDetailHeight);
+
+  apiRef.current.unstable_updateControlState({
+    stateId: 'detailPanels',
+    propModel: props.detailPanelExpandedRowIds,
+    propOnChange: props.onDetailPanelExpandedRowIdsChange,
+    stateSelector: gridExpandedRowIdsSelector,
+    changeEvent: GridEvents.detailPanelsExpandedRowIdsChange,
+  });
+
+  const toggleDetailPanel = React.useCallback<GridDetailPanelApi['toggleDetailPanel']>(
+    (id: GridRowId) => {
+      if (props.getDetailPanelContent == null) {
+        return;
+      }
+
+      const content = contentCache[id];
+      if (!React.isValidElement(content)) {
+        return;
+      }
+
+      setGridState((state) => {
```

Comment:
Can we call `apiRef.current.setExpandedRowIds` here ?
That way, if we add checks in `setExpandedRowIds`, all state updates will automatically have them.

Reply:
Done.